pkg/providers/vsphere/vmprovider_vm_snapshot.go

```diff
@@ -819,3 +828,29 @@ func (vs *vSphereVMProvider) reconcileCurrentSnapshot(
 
 	return nil
 }
+
+// unmarshalAndConvertVMFromYAML unmarshals VM YAML from snapshot and converts it
+// to the latest version of VirtualMachine.
+func (vs *vSphereVMProvider) unmarshalAndConvertVMFromYAML(
+	vmCtx pkgctx.VirtualMachineContext,
+	vmYAML string) (*vmopv1.VirtualMachine, error) {
+
+	decUnstructured := yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+
+	unstructuredObj := &unstructured.Unstructured{}
+	if _, _, err := decUnstructured.Decode([]byte(vmYAML), nil, unstructuredObj); err != nil {
+		vmCtx.Logger.Error(err, "failed to decode VM YAML in to Unstructured object")
+		return nil, err
+	}
+
+	// Convert it to newest version of VirtualMachine.
+	vm := &vmopv1.VirtualMachine{}
+	if err := vs.k8sClient.Scheme().Convert(unstructuredObj, vm, nil); err != nil {
+		vmCtx.Logger.Error(err, "failed to convert VM YAML to VirtualMachine",
+			"currentAPIVersion", unstructuredObj.GetAPIVersion(),
+			"desiredAPIVersion", vmopv1.GroupVersion.String())
+		return nil, err
+	}
```

Comment:
Should we just return the error? Since the error message will propagate up anyway.